File: ios/Cookpit/Cookpit/ExploreDataController.swift
//
//  ExploreDataController.swift
//  Cookpit
//
//  Created by Kittinun Vantasin on 7/7/16.
//  Copyright © 2016 Cookpit. All rights reserved.
//

import Foundation
import RxSwift

class ExploreDataController {

  private let controller = CPExploreController.create()!

  fileprivate let _viewData = Variable<CPExploreViewData?>(nil)

  lazy var viewData: Observable<CPExploreViewData> = {
    self._viewData.asObservable()
      .filter { $0 != nil }
      .map { $0! as CPExploreViewData }
      .distinctUntilChanged()
      .observeOn(MainScheduler.instance)
  }()
  
  fileprivate let _loadings = Variable(false)
  
  lazy var loadings: Observable<Bool> = {
    self._loadings.asObservable()
        .observeOn(MainScheduler.instance)
  }()
  
  lazy var errors: Observable<String> = {
    self._viewData.asObservable()
      .filter { $0 != nil && $0!.error }
      .distinctUntilChanged { $0!.error }
      .map { $0!.message }
      .observeOn(MainScheduler.instance)
  }()
  
  var currentPage: Int = 1
  
  init() {
    controller.subscribe(self)
  }
  
  func unsubscribe() {
    controller.unsubscribe()
  }
  
  func reset() {
    currentPage = 1
    controller.reset()
  }
  
  func request(page: Int) {
    currentPage = page
    
    self.controller.request(Int8(page))
  }

  func requestFromDb(page: Int) -> [CPExploreDetailViewData] {
    return controller.requestDb(Int8(page))
  }
  
  func requestNextPage() {
    if !_loadings.value {
        request(page: currentPage + 1)
    }
  }

  deinit {
    controller.unsubscribe()
  }

}

extension ExploreDataController : CPExploreControllerObserver {

  func onBeginUpdate() {
    _loadings.value = true
  }

  func onUpdate(_ viewData: CPExploreViewData) {
    _viewData.value = viewData
  }

  func onEndUpdate() {
    _loadings.value = false
  }

}

File: android/Cookpit/app/src/main/kotlin/com/github/kittinunf/cookpit/explore/ExploreDataController.kt
package com.github.kittinunf.cookpit.explore

import com.github.kittinunf.cookpit.ExploreController
import com.github.kittinunf.cookpit.ExploreControllerObserver
import com.github.kittinunf.cookpit.ExploreDetailViewData
import com.github.kittinunf.cookpit.ExploreViewData
import com.github.kittinunf.reactiveandroid.MutableProperty

class ExploreDataController : ExploreControllerObserver() {

    private val controller = ExploreController.create()

    private val _viewData = MutableProperty<ExploreViewData>()

    val viewData by lazy { _viewData.observable }

    private val _loadings = MutableProperty(false)

    val loadings by lazy { _loadings.observable }

    private val _loadingMores = MutableProperty(false)

    val loadingMores by lazy { _loadingMores.observable }

    val errors by lazy {
        _viewData.observable.filter { it.error }
                .distinctUntilChanged()
                .map { it.message }
    }

    var currentPage = 1

    init {
        controller.subscribe(this)
    }

    fun reset() {
        currentPage = 1
        controller.reset()
    }

    fun request(page: Int) {
        currentPage = page
        controller.request(page.toByte())
    }

    fun requestDb(page: Int): List<ExploreDetailViewData> = controller.requestDb(page.toByte())

    fun requestNextPage() {
        request(currentPage + 1)
    }

    fun unsubscribe() {
        controller.unsubscribe()
    }

    //region ExploreControllerObserver
    override fun onBeginUpdate() {
        if (currentPage == 1) {
            _loadings.value = true
        } else {
            _loadingMores.value = true
        }
    }

    override fun onUpdate(viewData: ExploreViewData?) {
        _viewData.value = viewData
    }

    override fun onEndUpdate() {
        _loadings.value = false
        _loadingMores.value = false
    }
    //endregion

}